### DISCUSSION TITLE
maturin fails to package .so.1 in target directory

### DISCUSSION THREAD

**Discussion by jdiggans-twist** [PyO3/maturin#1087] [Q&A] [answered]
I have a PyO3 project with python and Rust code. One of the Rust crate dependencies is [catboost](https://github.com/catboost/catboost), which has its own [build.rs script](https://github.com/catboost/catboost/blob/master/catboost/rust-package/catboost-sys/build.rs) and compiles an rlib and an .so.1 library file. 

The catboost .so.1 file *is* correctly build by maturin (triggering cargo build) and is present deep in a `target/build` subdirectory. Adding this via `LD_LIBRARY_PATH` does not seem to help maturin find the file. I am using quay.io/pypa/manylinux_2_28_x86_64 as my base image, so everything around the manylinux stuff seems to work fine.

When trying to build a manylinux_2_28 wheel using maturin, however, I get the following:
```
💥 maturin failed
 Caused by: Cannot repair wheel, because required library libcatboostmodel.so.1 could not be located.
```

I can build the wheel using `--skip-auditwheel` and then try repairing but again get failure:

```
ValueError: Cannot repair wheel, because required library "libcatboostmodel.so.1" could not be located
```

If I unzip the wheel and look at the output of `ldd` on the binary I see:

```
linux-vdso.so.1 (0x00007ffc6d1a8000)
libcatboostmodel.so.1 => not found
libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f1a413eb000)
libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f1a411cb000)
libm.so.6 => /lib64/libm.so.6 (0x00007f1a40e49000)
libdl.so.2 => /lib64/libdl.so.2 (0x00007f1a40c45000)
libc.so.6 => /lib64/libc.so.6 (0x00007f1a40880000)
/lib64/ld-linux-x86-64.so.2 (0x00007f1a41cfc000)
```

How can I help maturin find and include this file? Why is it ignoring the library?

**Comment by messense** [comment, marked as answer]
Both upstream and our auditwheel did not take `.so` files in `target` directory into consideration thus the error.

https://github.com/PyO3/maturin/blob/2934c6770e2b6f3de8cbb159f00a7b0396c536cc/src/auditwheel/repair.rs#L12

`LD_LIBRARY_PATH` only works when not cross compiling, but it could be a bug if it does not work for you, I didn't test it thoroughly.

https://github.com/messense/lddtree-rs/blob/9ea2b0cedff62f199f837b6b0e9f4449e301b236/src/lib.rs#L196-L200

Edit: Fixed in https://github.com/PyO3/maturin/pull/1092

**Reply by messense** [reply]
IMO, maturin should make it work since it's a common practice in Rust to put build artifacts in target directory, so I'm going to turn this into an issue. Let's continue the discussion there: https://github.com/PyO3/maturin/issues/1088
